api/src/main/java/me/lokka30/treasury/api/economy/currency/Currency.java

```diff
@@ -12,69 +12,216 @@
  * This allows economy providers and plugins to use different
  * currencies for different transactions, e.g. a daily reward
  * plugin can award players 'Tokens', but a job plugin
  * can award players 'Dollars'. Facilitates great customisability.
+ *
+ * @author lokka30, Geolykt, MrIvanPlays
+ * @since v1.0.0
  */
-@SuppressWarnings("unused")
-public interface Currency {
+public class Currency {
+
+    /**
+     * Creates a new currency with random {@link UUID} {@code currencyId} and a {@code startingBalance} handler of
+     * always returning 0.
+     *
+     * @param currencyChar          a way to identify the currency via a character e.g. '$' for a dollar. can be null.
+     * @param roundedDigits         to which digit shall we round balances. can specify -1 for no rounding.
+     * @param conversionCoefficient coefficient at which this currency is going to be converted to other currencies.
+     * @param balanceFormatter      a function, giving a human-readable format of a balance.
+     * @param names                 acceptable names of the currency. primary name shall be the first specified.
+     * @return new currency instance
+     */
+    @NotNull
+    public static Currency of(
+            @Nullable Character currencyChar,
+            int roundedDigits,
+            double conversionCoefficient,
+            @NotNull BiFunction<Double, Locale, String> balanceFormatter,
+            @NotNull String... names
+    ) {
+        return new Currency(
+                null,
+                currencyChar,
+                roundedDigits,
+                conversionCoefficient,
+                null,
+                balanceFormatter,
+                names
+        );
+    }
+
+    /**
+     * Creates a new currency.
+     *
+     * @param currencyId            the currency id.
+     * @param currencyChar          a way to identify the currency via a character e.g. '$' for a dollar. can be null.
+     * @param roundedDigits         to which digit shall we round balances. can specify -1 for no rounding.
+     * @param conversionCoefficient coefficient at which this currency is going to be converted to other currencies.
+     * @param startingBalance       a function, giving a starting balance for the inputted player uuid, which can be null.
+     * @param balanceFormatter      a function, giving a human-readable format of a balance.
+     * @param names                 acceptable names of the currency. primary name shall be the first specified.
+     * @return new currency instance
+     */
+    @NotNull
+    public static Currency of(
+            @NotNull UUID currencyId,
+            @Nullable Character currencyChar,
+            int roundedDigits,
+            double conversionCoefficient,
+            @Nullable Function<UUID, Double> startingBalance,
+            @NotNull BiFunction<Double, Locale, String> balanceFormatter,
+            @NotNull String... names
+    ) {
+        Objects.requireNonNull(currencyId, "currencyId");
```

Comment:
Aren't all of these strictly redundant? I'm almost certain the JetBrains `@NotNull` annotation is processed during compile to include the required validations.

Reply:
no. jetbrains' annotations are only a javadoc perk and a perk for intellij users so it bugs them to check nullability.

Reply:
https://www.jetbrains.com/help/idea/nullable-and-notnull-annotations.html

> When you compile your project, the IDE adds assertions to all methods and parameters annotated with the `@NotNull` annotation. The assertions will fail if null is passed in code where `@NotNull` is expected. You can disable this option and configure the list of annotations in the Settings/Preferences dialog Ctrl+Alt+S. Go to Build, Execution, Deployment | Compiler.

Looks like it's not just syntactical sugar to me, but I'm not sure if that's purely on IntelliJ's side. I know for a fact I've had the checks generate while compiling with Maven, however, that was inside IntelliJ.

Reply:
Just compiled via the terminal:
![decompiler-source](https://img.mrivanplays.com/xhufksiljt.png)

do you see anything generated? I may be blind or idk, but I don't see it.
Irony aside, the plugin is going to be build and deployed to maven repositories via a CI and not by intellij/human, so if you want that code to generate not just if you compile via intellij, you need to find a maven plugin which does it and put it. But that has nothing to do with this PR.

Reply:
Yeah, looks like it's IJ-specific. I'll have to look into it more, would be useful to have.